Swift 5.7 compiler error lacks file name or line number when block doesn't have the right return type

**Describe the bug**

The following code should (and does) fail to compile, but the error message is a lot less helpful under swift 5.7

```
["foo"].map { s in
    if s == "1" { return }
    return s.isEmpty
}.filter { $0 }
```

**To Reproduce**

Compiled with swift5.7, the error message doesn't list a file name or line number, so it's almost impossible to find the bug.

```
$ swift ~/Desktop/error.swift 
<unknown>:0: error: cannot convert return expression of type '()' to return type 'Bool'
```

**Expected behavior**

Compiled with swift5.6, the error message is much more useful.

```
$ swift5.6 ~/Desktop/error.swift 
/Users/tominsam/Desktop/error.swift:2:19: error: non-void function should return a value
    if s == "1" { return }
                  ^
```


**Environment (please complete the following information):**
 - OS: macOS 12.4 (M1)
 - swift 5.7 from xcode 14b2 (14A5229c)
 - swift 5.6 from xcode 13.4 (13F17a)

--- a/lib/Sema/CSClosure.cpp
+++ b/lib/Sema/CSClosure.cpp
@@ -26,10 +26,10 @@ using namespace swift::constraints;
 namespace {
 
 // Produce an implicit empty tuple expression.
-Expr *getVoidExpr(ASTContext &ctx) {
+Expr *getVoidExpr(ASTContext &ctx, SourceLoc contextLoc = SourceLoc()) {
   auto *voidExpr = TupleExpr::createEmpty(ctx,
-                                          /*LParenLoc=*/SourceLoc(),
-                                          /*RParenLoc=*/SourceLoc(),
+                                          /*LParenLoc=*/contextLoc,
+                                          /*RParenLoc=*/contextLoc,
                                           /*Implicit=*/true);
   voidExpr->setType(ctx.TheEmptyTupleType);
   return voidExpr;
@@ -885,11 +885,13 @@ private:
     } else {
       // If this is simplify `return`, let's create an empty tuple
       // which is also useful if contextual turns out to be e.g. `Void?`.
-      resultExpr = getVoidExpr(cs.getASTContext());
+      // Also, attach return stmt source location so if there is a contextual
+      // mismatch we can produce a diagnostic in a valid source location.
+      resultExpr = getVoidExpr(cs.getASTContext(), returnStmt->getEndLoc());
     }
 
     SolutionApplicationTarget target(resultExpr, context.getAsDeclContext(),
-                                     CTP_ReturnStmt, resultType,
+                                     CTP_ClosureResult, resultType,
                                      /*isDiscarded=*/false);
 
     if (cs.generateConstraints(target, FreeTypeVariableBinding::Disallow)) {
@@ -898,7 +900,7 @@ private:
     }
 
     cs.setContextualType(target.getAsExpr(), TypeLoc::withoutLoc(resultType),
-                         CTP_ReturnStmt);
+                         CTP_ClosureResult);
     cs.setSolutionApplicationTarget(returnStmt, target);
   }
 
@@ -1508,7 +1510,7 @@ private:
     // number of times.
     {
       SolutionApplicationTarget target(resultExpr, context.getAsDeclContext(),
-                                       CTP_ReturnStmt, resultType,
+                                       CTP_ClosureResult, resultType,
                                        /*isDiscarded=*/false);
       cs.setSolutionApplicationTarget(returnStmt, target);
 
@@ -1574,7 +1576,7 @@ private:
 
     SolutionApplicationTarget resultTarget(
         resultExpr, context.getAsDeclContext(),
-        mode == convertToResult ? CTP_ReturnStmt : CTP_Unused,
+        mode == convertToResult ? CTP_ClosureResult : CTP_Unused,
         mode == convertToResult ? resultType : Type(),
         /*isDiscarded=*/false);
     if (auto newResultTarget = rewriteTarget(resultTarget))

--- a/test/Constraints/callAsFunction.swift
+++ b/test/Constraints/callAsFunction.swift
@@ -83,7 +83,7 @@ struct Test {
   var body8: MyLayout {
     MyLayout {
       let x = ""
-      return x // expected-error {{cannot convert return expression of type 'String' to return type 'Int'}}
+      return x // expected-error {{cannot convert value of type 'String' to closure result type 'Int'}}
     } content: {
       EmptyView()
     }

--- a/test/decl/circularity.swift
+++ b/test/decl/circularity.swift
@@ -41,7 +41,7 @@ class Sub: Base {
     var foo = { () -> Int in
         let x = 42
         // FIXME: Bogus diagnostic
-        return foo(1) // expected-error {{cannot convert return expression of type '()' to return type 'Int'}}
+        return foo(1) // expected-error {{cannot convert value of type '()' to closure result type 'Int'}}
     }()
 }
 

--- a/test/expr/closure/closures.swift
+++ b/test/expr/closure/closures.swift
@@ -690,7 +690,7 @@ func testSR13239_ArgsFn() -> Int {
 func testSR13239MultiExpr() -> Int {
   callit {
     print("hello") 
-    return print("hello") // expected-error {{cannot convert return expression of type '()' to return type 'Int'}}
+    return print("hello") // expected-error {{cannot convert value of type '()' to closure result type 'Int'}}
   }
 }
 
@@ -732,3 +732,21 @@ public class TestImplicitCaptureOfExplicitCaptureOfSelfInEscapingClosure {
         }
     }
 }
+
+// https://github.com/apple/swift/issues/59716
+["foo"].map { s in
+    if s == "1" { return } // expected-error{{cannot convert value of type '()' to closure result type 'Bool'}}
+    return s.isEmpty
+}.filter { $0 }
+
+["foo"].map { s in
+    if s == "1" { return } // expected-error{{cannot convert value of type '()' to closure result type 'Bool'}}
+    if s == "2" { return }
+    if s == "3" { return }
+    return s.isEmpty
+}.filter { $0 }
+
+["foo"].map { s in
+    if s == "1" { return () } // expected-error{{cannot convert value of type '()' to closure result type 'Bool'}}
+    return s.isEmpty
+}.filter { $0 }

